libglusterfs/src/inode.c

```diff
@@ -450,16 +450,15 @@ __inode_retire(inode_t *inode)
 static int
 __inode_get_xl_index(inode_t *inode, xlator_t *xlator)
 {
-    int set_idx = -1;
+    int set_idx = xlator->xl_id;
+    xlator_t *xl_key = inode->_ctx[set_idx].xl_key;
 
-    if ((inode->_ctx[xlator->xl_id].xl_key != NULL) &&
-        (inode->_ctx[xlator->xl_id].xl_key != xlator))
-        goto out;
+    if (xl_key == NULL)
+        xl_key = xlator;
+    else if (caa_unlikely(xl_key != xlator))
+        set_idx = -1;
 
-    set_idx = xlator->xl_id;
-    inode->_ctx[set_idx].xl_key = xlator;
-
-out:
+    /* Returns -1 when xl_key != NULL && xl_key != xlator */
```

Comment:
Given that currently `xlator->xl_id` is a unique value in the entire graph, I wonder how `xl_key` could have another value. Looking at the line [coverage test](https://build.gluster.org/job/line-coverage/Line_20Coverage_20Report/) I see that this happens in 0.00003% of the cases. How's that possible ? is it a bug ? or am I missing some case ?

I would prefer to identify those corner cases and fix them to be able to completely get rid of the `xl_key` field.

Reply:
1. Let's crash when it happens, so we can figure that one out? (or just save a backtrace...)
2. How will this work with Mohit's changes around this function?

Reply:
> 1. Let's crash when it happens, so we can figure that one out? (or just save a backtrace...)

I'll try that.

> 2. How will this work with Mohit's changes around this function?

I'm still reviewing @mohit84 changes, but I think it shouldn't matter much. Even if the `xl_id` is not globally unique, it should still be unique in any given branch, otherwise some xlator wouldn't be able to associate data to the inode, which would break it. So we cannot have two xlators using the same entry.